Testcontainers guide blog (#3622)
OpenLiberty/blogs: posts/2024-03-13-testcontainers-guide.adoc

---
layout: post

> title: "Build true-to-production tests with Testcontainers, JUnit and Open Liberty"

Reviewer: ```suggestion
title: "Build true-to-production tests with Testcontainers, JUnit, and Open Liberty"
```

# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/GraceJansen
author_github: https://github.com/GraceJansen

> seo-title: Build true-to-production tests with Testcontainers, JUnit and Open Liberty - OpenLiberty.io

Reviewer: ```suggestion
seo-title: Build true-to-production tests with Testcontainers, JUnit, and Open Liberty - OpenLiberty.io
```

> seo-description: Find out more about Testcontainers, explore how they can enable true-to-production integration tests, and discover how you make use of them in your own cloud-native Java applications with our new interactive guide.

Reviewer: ```suggestion
seo-description: Explore how the Testcontainers framework enables true-to-production integration tests and discover how to use it in your own cloud-native Java applications with our new interactive guide.
```

> blog_description: "Find out more about Testcontainers, explore how they can enable true-to-production integration tests, and discover how you make use of them in your own cloud-native Java applications with our new interactive guide."

Reviewer: ```suggestion
blog_description: "Explore how the Testcontainers framework enables true-to-production integration tests and discover how to use it in your own cloud-native Java applications with our new interactive guide."
```

open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
---
= Build true-to-production tests with Testcontainers, JUnit and Open Liberty
Grace Jansen <https://github.com/GraceJansen>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

> == Why is there a need for true-to-production tests?

Reviewer: ```suggestion
== Why do we need true-to-production tests?
```

> In a world where our cloud-native Java applications are often deployed to the cloud with the use of containers, it makes sense that we should be testing our applications in containers as well. As the link:https://developer.ibm.com/articles/creating-a-12-factor-application-with-open-liberty/[12 factor] and link:https://developer.ibm.com/articles/15-factor-applications/[15 factor] methodologies promote, ensuring parity between our deployment environments is critical for us to have confidence in our application's expected behaviour and spotting/solving bugs. 

Reviewer: ```suggestion
In a world where our cloud-native Java applications are often deployed to the cloud in containers, it makes sense that we should also be testing these applications in containers. As the link:https://developer.ibm.com/articles/creating-a-12-factor-application-with-open-liberty/[12 factor] and link:https://developer.ibm.com/articles/15-factor-applications/[15 factor] methodologies promote, ensuring parity between our deployment environments builds confidence in our application's expected behaviour and helps us spot and solve bugs. 
```

image::/img/blog/environmentalParity.png[Image showcasing environmental parity,width=50%,align="center"]

> Containers are an important part of this parity. One of the great benefits of containers (and container technologies like Docker) is that we get a nice consistent package that contains everything we need, down to the OS, meaning it’s portable to any hardware. So why not use these technologies to obtain these benefits when it comes to testing our applications and ensure we get consistent testing environments too?!

Reviewer: ```suggestion
Containers are an important part of this parity. One of the great benefits of containers (and container technologies like Docker) is that we get a nice consistent package that contains everything we need, down to the OS, that is portable to any hardware. So why not use these technologies to ensure we get consistent testing environments too?!
```

How, then, can we effectively utilise containers in our tests and ensure we're testing our applications in a manner that is as true to production as possible? This is where the Testcontainers framework can help.

== What is the Testcontainers framework?

link:https://testcontainers.com/[Testcontainers] is an open source library that provides containers as a resource at test time, creating consistent and portable testing environments. Breaking it down, it essentially wraps Docker in a Java API and offers first-class integration with JUnit. 

image::/img/blog/testcontainers_banner.png[Testcontainers logo,width=50%,align="center"]

== Why is Testcontainers helpful?

> Although this technology is especially useful for applications that have external resource dependencies (such as databases, message queues, or web services), one of the most important features of Testcontainers is the generic support it has for any Docker image. By encapsulating any necessary dependencies in containers, Testcontainers simplifies the configuration process and ensures a uniform testing setup that closely mirrors production environments.

Reviewer: ```suggestion
Although this framework is especially useful for applications that have external resource dependencies (such as databases, message queues, or web services), one of the most important features of Testcontainers is generic support for any Docker image. By encapsulating any necessary dependencies in containers, Testcontainers simplifies the configuration process and ensures a uniform testing setup that closely mirrors production environments.
```

image::/img/blog/TestContainers3.png[Diagram of generic Testcontainers architecture,width=50%,align="center"]

In essence, with Testcontainers, you can run your tests in an isolated and controlled environment that closely resembles your production setup, ensuring that your tests are reliable and reproducible.

As well as this increased reliability and reproducibility of tests, there are additional key features/benefits that make Testcontainers a really useful tool, including:

> * Offering a wide range of supported containers

Reviewer: ```suggestion
* A wide range of supported containers
```

+

> Testcontainers offers support for a large variety of different Docker containers, including databases (e.g., PostgreSQL, MySQL, MongoDB, DB2, Redis), messaging brokers (e.g., RabbitMQ, Kafka), web servers, and more. This diversity enables developers to be able to customise test environments to make them specific to their own application stack.

Reviewer: ```suggestion
Testcontainers supports a large variety of Docker containers, including databases (e.g., PostgreSQL, MySQL, MongoDB, DB2, Redis), messaging brokers (e.g., RabbitMQ, Kafka), web servers, and more. This diversity enables you to customise test environments to make them specific to your application stack.
```

* Integration with popular testing frameworks

+

> Testcontainers integrates with popular testing frameworks including JUnit, TestNG, and Spock. This enables developers to easily and efficiently incorporate Testcontainers into their testing suite without having to make significant changes to their existing tests.

Reviewer: ```suggestion
Testcontainers integrates with popular testing frameworks including JUnit, TestNG, and Spock. You can easily and efficiently incorporate Testcontainers into your testing suite without having to make significant changes to your existing tests.
```

* Declarative configuration

+

> Testcontainers provides a simple, declarative API for configuring and deploying containers. Developers are able to define any desired properties of a container using code, making it easy to configure and maintain.

Reviewer: ```suggestion
Testcontainers provides a simple, declarative API for configuring and deploying containers. You can define any desired properties of a container using code, making it easy to configure and maintain.
```

* Lifecycle management

+

> Testcontainers also handles the lifecycle of containers, ensuring they are started before the tests run and, importantly, stopped afterward. This automation removes this responsibility from the shoulders of developers, eliminating the need for manual intervention and making it easy to maintain clean, isolated testing environments.

Reviewer: ```suggestion
Testcontainers also handles the lifecycle of containers, ensuring they are started before the tests run and, importantly, stopped afterward. This automation removes this responsibility from your shoulders, eliminating the need for manual intervention and making it easy to maintain clean, isolated testing environments.
```

* Wait strategies

+

> To prevent your tests from trying to run before any required containers are set up, Testcontainers has a helpful, built-in wait strategy. This is of particular importance for any services that have tests that rely on services that can take significant time to initialise (e.g. databases).

Reviewer: ```suggestion
To prevent your tests from trying to run before any required containers are set up, Testcontainers has a helpful, built-in wait strategy. This particularly important for services with tests that rely on services that can take significant time to initialise (e.g. databases).
```

> == How to make use of this in your own applications

Reviewer: ```suggestion
== How to use Testcontainers in your own applications
```

In our newest guide, we walk you through how you can add Testcontainers to an existing Java application. This interactive guide shows how to to set up and configure multiple containers, including the Open Liberty Docker container, to simulate a production-like environment for your tests. The guide uses Docker to run an instance of the PostgreSQL database needed in this application for a fast installation and setup.

image::/img/blog/guideArchitecture.png[Architecture of the Testcontainers guide,width=50%,align="center"]

> As Testcontainers is a black-box form of testing, you will need a REST client to be able to trigger the requests for your tests. In this guide we also show how to build a test REST client to accurately verify the application’s behavior by ensuring that it responds correctly to various scenarios and conditions.

Reviewer: ```suggestion
As Testcontainers is a black-box form of testing, you  need a REST client to trigger the requests for your tests. In this guide, we also show how to build a test REST client to accurately verify the application’s behavior by ensuring that it responds correctly to various scenarios and conditions.
```
